[2단계 - 경로 조회 기능] 바론(이소민) 미션 제출합니다.

somsom13 opened the pull request:
안녕하세요 영이!!! 바론입니다 🙇‍♀️

2단계에서는 최단 경로 조회와 요금 계산 로직을 구현했습니다.

그리고 지난번에 남겨주셨던 리뷰들을 반영하고, SectionEntity VS SectionDto 질문에 코멘트 답변을 남겼습니다!

[https://github.com/woowacourse/jwp-subway-path/pull/29#discussion_r1195257547](https://github.com/woowacourse/jwp-subway-path/pull/29#discussion_r1195257547)

---

미션을 진행하면서 한 가지 고민이 되었던 점은…

**GET메서드에 body를 넣어도 되는걸까?** 입니다.

HTTP 공식 문서에서는 Get에 body를 넣는 것이 가능은 하지만, 서버에서 get의 body를 처리할 수 있는 기능이 없다면 그냥 무시되는 값이라고 합니다.

저는 그래서 GET요청에는 body를 넣지 않는게 보편적으로 맞는 방법이라고 생각했는데요,

이번에 구현한 경로 조회 API는 “최단 경로를 알려줘!”이기 때문에 의미적으로는 GET이 맞는 것 같지만, 요청을 받아올 때 RequestBody로 받아오도록 했습니다.

- RequestBody로 받아오는 방법과 (빈 값, null 검증 가능)
- queryParam으로 받아오는 방법

두 가지 방법이 떠오르는데요, 혹시 영이라면 어떤 방법을 선택하실 것 같은지 의견을 여쭤보고 싶습니다!

---

### 프로덕션 데이터베이스 docker 실행
```text
# 루트 디렉터리에서 실행 (jwp-subway-path)
cd docker
docker-compose -p subway up -d
```

---

이번 리뷰도 잘 부탁드립니다. 감사합니다 😻

somsom13 committed:
docs: API 명세 수정 및 id 반환값 추가

somsom13 committed:
refactor: Line 반환값 필드명 변경

somsom13 committed:
refactor: 테스트 URI 변경

somsom13 committed:
refactor: LineService, Controller DTO 분리

somsom13 committed:
refactor: SectionService, Controller DTO 분리

somsom13 committed:
refactor: StationService, StationController DTO 분리

somsom13 committed:
docs: 삭제된 API 명세 제거

somsom13 committed:
docs: 최단 경로 및 요금 계산 기능 요구사항 작성

somsom13 committed:
feat: jgrapht를 사용하여 최단 경로 조회 도메인 생성

somsom13 committed:
feat: 출발역에서 도착역까지 최단 경로 조회 및 거리 반환

somsom13 committed:
feat: 경로 길이에 따른 요금 계산

somsom13 committed:
refactor: Section 조회 메서드명 변경

somsom13 committed:
feat: 최단 경로 조회 API 구현

somsom13 committed:
feat: 프로덕션, 테스트 데이터베이스 분리

somsom13 committed:
feat: 예외 문자열 response 포장

somsom13 committed:
refactor: 새 구간역 추가 후 추가된 구간 반환하도록 변경

somsom13 committed:
refactor: dto 패키지 controller, service 레이어로 분리

choijy1705 commented on src/main/java/subway/ui/GlobalExceptionHandler.java:
```diff
@@ -22,27 +24,28 @@ public class GlobalExceptionHandler {
     @ExceptionHandler(MethodArgumentNotValidException.class)
-    public ResponseEntity<Map<String, String>> handleMethodArgumentNotValidException(MethodArgumentNotValidException e) {
+    public ResponseEntity<ExceptionResponse> handleMethodArgumentNotValidException(MethodArgumentNotValidException e) {
         Map<String, String> validation = new HashMap<>();
```
요건 안쓰이는거 같네요

somsom13 replied:
세상에.. 이번에야 말로 사용되지 않는 메서드, 코드는 모두 제거하고 리뷰 요청 드려야지! 라고 마음 먹었는데,,,, ㅠㅠ 

제거해두고 이걸 푸쉬를 하지 않았었네요 😿 이런 바보같은 실수를 하다니.. 

다음 리뷰 요청 드릴 때 반영하겠습니다!! 🙇‍♀️

choijy1705 commented on src/main/java/subway/ui/PathController.java:
```diff
@@ -0,0 +1,28 @@
+
+@RestController
+@RequestMapping("/routes")
+public class PathController {
```
uri는 routes 인데 controller는 path라서 혼동을 줄 여지가 있어보이네요!

somsom13 replied:
맞아요 😂 😂 
네이밍을 하고 나서도 고민을 많이 했는데, 괜찮지 않을까.. 라는 마음으로 그냥 넘겨버렸네요 ㅎㅎ...
둘 중에 현재 코드에서 더 많이 사용하고 있는 용어로 통일하겠습니다!

choijy1705 commented on src/main/java/subway/ui/PathController.java:
```diff
@@ -0,0 +1,28 @@
+    }
+
+    @GetMapping
+    public ResponseEntity<ShortestPathResponse> findRoute(@RequestBody @Valid PathFindRequest request) {
```
get 요청에서는 body를 사용하기보다는 param을 사용하는게 좋을것 같습니다.
물론 get 요청에 body를 넣어줘도 사용할 수 있지만
rest의 하나의 약속으로서 일반적으로 get은 당연히 param으로 사용한다라고 생각할것같아요!
`@RequestParam` 어노테이션에서 required옵션을 통하여 null 체크정도는 할 수 있을것 같네요 
<img width="717" alt="image" src="https://github.com/woowacourse/jwp-subway-path/assets/63634505/04eacdf7-cea7-4f1d-b8e7-c5af95ed7557">

choijy1705 replied:
그리고 name을 통하여 조회를 하는것에 대하여도 고민해보면 좋겠네요!
id를 통해 조회하는것과 name을 통해 조회하는것은 어떤 차이가 있을까요?

somsom13 replied:
우선 name을 사용한 이유는 "사용자가 출발역, 도착역의 이름을 직접 입력한다고 생각했기 때문" 이었습니다!

그런데 생각해보니, 대부분의 지하철 노선도 어플리케이션을 보면 노선도에서 출발역과 도착역을 선택하는 경우가 더 많았던 것 같습니다. 이런 흐름으로 생각하면 id로 받아오는게 맞을 것 같습니다.

- id는 한 객체를 식별하는 고유한 식별자이기 때문에, 클라이언트에서 완전하게 잘못된 값을 넘기지 않는 이상은 문제가 생길 가능성이 적다는 장점이 있는 반면,
- name을 직접 받는 것은 1단계 리뷰에서도 말씀해주셨던 것 처럼 엣지 케이스가 많이 발생할 것 같습니다. (비정상적인 문자의 포함, 공백이나 개행문자 포함 등으로 인한 문제 등등요!)

id로 입력을 받는 방향이 더 적합하다는 생각이 들어서 해당 방식으로 변경하겠습니다. 감사합니다 :)

choijy1705 replied:
바론이 말씀해주신 내용도 이유가 될 수 있지만
name같은 경우는 index가 따로 설정되어 있지 않아서 
name을 통해서 조회시 성능이 느려질 우려가 있어서 말씀드렸습니다!
unique 제약조건만 걸려있는것 같네요

choijy1705 commented on src/main/java/subway/domain/FareCalculator.java:
```diff
@@ -0,0 +1,47 @@
+
+import org.springframework.stereotype.Component;
+
+@Component
```
`@Componet`를 통하여 빈등록을 하신 이유가 궁금하네요!
또한 FareCalculator는 util 클래스에 좀 더 가까운것 같은데 스프링 빈으로 관리될 필요가 있을까요?

somsom13 replied:
"요금 계산 정책의 변경 가능성" 이 높다고 생각했기 때문이었습니다..!
(실제로도 지하철 요금이 변동되는 경우가 많다고 생각했기 때문입니다!)

이러한 변경 가능성을 생각한다면 Calculator를 인터페이스로 분리하고, 구현체를 런타임에 끼워주는 방식이 적합할 것이라 생각했습니다.
Spring이 관리하는 DI를 위해서라면, 당연히 Calculator를 스프링 빈으로 등록해야 할 것이라 생각했습니다!

말씀해주신 것 처럼 util성 클래스이기 때문에 사실 static 메서드로 두어도 되는게 아닐까? 라는 생각도 하고 있습니다. 😂 

혹시 영이의 의견도 여쭤보아도 괜찮을까요??

choijy1705 replied:
util 클래스는 static 메서드로 관리하는게 좋을것 같네요
혹시나 상태를 가지는 필드를 누군가 추가하게 된다면
util클래스가 상태정보를 가져버려 문제가 생길 가능성이 있을수도 있을것 같네요
가능하면 util 클래스는 인스턴스화 되지 않도록 하는게 좋을것 같아요

choijy1705 commented on src/main/java/subway/domain/FareCalculator.java:
```diff
@@ -0,0 +1,47 @@
+    private static final int FIFTY = 50;
+
+
+    public int calculate(Path path) {
```
돈과 관련된 연산은 BigDecimal을 활용하는게 좋을것 같아요!

choijy1705 commented on src/main/java/subway/domain/FareCalculator.java:
```diff
@@ -0,0 +1,47 @@
+    private static final int OVER_10_FARE = 100;
+    private static final int OVER_50_FARE = 100;
+    private static final int OVER_10_CHARGE_UNIT = 5;
+    private static final int OVER_50_CHARGE_UNIT = 8;
```
OVER_10KM_FARE 는 어떨까요?
상수명만 봤을땐 어떤 숫자인지 느낌이 오지않네요!

choijy1705 commented on src/main/java/subway/domain/Line.java:
```diff
@@ -54,10 +54,14 @@ public Long getId() {
     }
 
-    public List<Section> getSections() {
+    public List<Section> getSectionsByList() {
```
바꾸신 이유가 있나요?
getSections가 더 좋은것 같은데. 변수명, 메서드 명이든 자료구조가 나타나는건 그렇게 좋은것 같지 않아요

somsom13 replied:
저도 변수명, 메서드명에 자료구조가 나타나지 않아야 한다는 의견에 완전히 동의합니다 😂 

그런데, 구현을 하다보니 `Sections` 객체를 완전히 꺼내오는 것과 `List<Section>` 을 꺼내오는 메서드가 두 가지 다 필요해지길래 해당 네이밍을 선택했었습니다...ㅎㅎ 

방금 usage를 확인해보니 `getSections()`는 한 곳에서 밖에 사용이 안되어서 하나의 메서드로 통일할 수 있을 것 같아요! 감사합니다 :)

choijy1705 commented on src/main/java/subway/domain/PathFinder.java:
```diff
@@ -0,0 +1,64 @@
+import subway.exception.IllegalStationException;
+import subway.exception.StationNotFoundException;
+
+public class PathFinder {
```
한줄에 100자 넘는 부분들이 많은데
100자를 안넘도록 하면 좋을것 같아요🙏

somsom13 replied:
우테코 코드스타일을 적용해 두었는데, 지금 보니 Hard Wrap이 기본값인 120으로 설정이 되어있었네요 ㅠ.ㅠ

혹시 100이 표준적으로 더 많이 사용되는 값일까요??

choijy1705 replied:
음 회사마다 다를것 같은데
보통은 구글 컨벤션을 많이 따르는것 같아요
구글 컨벤션은 100자로 알고 있습니다

choijy1705 commented on src/main/java/subway/domain/PathFinder.java:
```diff
@@ -0,0 +1,64 @@
+        for (Sections sections : totalSections) {
+            addSectionsToGraph(sections.getSections(), graph);
+        }
+    }
```
불필요한 메서드 추출이지 않을까 싶네요

choijy1705 commented on src/main/java/subway/application/PathService.java:
```diff
@@ -0,0 +1,49 @@
+        return new ShortestPathDto(path, fare);
+    }
+
+    private List<Sections> findAllSections() {
```
section만 조회하면 될것 같은데 불필요하게 line까지 조회를 하는게 건 아닌가요?

choijy1705 commented on src/main/java/subway/application/SectionService.java:
```diff
@@ -26,37 +26,37 @@ public class SectionService {
+        Station endStation = stationRepository.findByName(requestedSection.getEndStationName());
+        Line line = lineRepository.findById(requestedSection.getLineId());
+        return addSection(line, new Section(startStation, endStation, requestedSection.getDistance()));
     }
```
👍

choijy1705 commented on src/main/resources/application.yml:
```diff
@@ -1,7 +1,10 @@
+  sql:
+    init:
+      data-locations: classpath:/schema.sql
+      mode: always
```
지금은 main 모듈, test 모듈에서 각각 application.yml 을 관리하고 있는데요
yml 파일 같은 경우 한곳에 모여져서 관리되는것이 좋다고 생각합니다!
profile을 통해서 main module에서 관리되도록 해보면 어떨까요?

choijy1705 commented on src/main/java/subway/application/dto/line/LineCreateDto.java:
```diff
@@ -0,0 +1,19 @@
+package subway.application.dto.line;
+
+public class LineCreateDto {
```
LineCreateRequest, LineCreateDto 는 사실상 똑같은 필드를 가지고 있고 필드수도 적은데 
저라면 굳이 dto를 만들지는 않을것 같아요
클래스의 숫자만 많아지고
dto의 사용의 목적중 하나가 변경을 줄일수 있다라는 점인데 이렇게 사용하면 request가 변경되면 오히려 변경해야할 부분이 더 많지는 않을까요?
어떤 방법이 좋다라는 것이 아니라 바론의 생각을 듣고 싶어서요!🙏
dto를 사용했을때 현재 구조에서는 어떤 장점이 있나요?

somsom13 replied:
우선 Service의 DTO와 Controller의 Request, Response를 분리한 이유는

서비스에서 Response를 만들지 않고 Controller에서 Response로 매핑하는 것에 대한 1단계의 피드백을 적용하면서, (https://github.com/woowacourse/jwp-subway-path/pull/29#discussion_r1195237845) "그러면 Request도 Service가 받는 DTO와 분리하는 것이 맞지 않을까?" 라는 생각을 했기 때문이었습니다!

지난번 미션의 리뷰에서도 "Controller와 Service의 DTO 분리"에 대한 피드백을 받았었는데, 1단계 때 영이가 남겨주신 리뷰와 합쳐서 이런 결론을 내렸습니다.

Request나 Response 요구사항이 달라지는 상황에서
1. Request가 다른 방식으로 넘어온다면? (사용자로부터 입력받는 요구사항이 달라지는 등) Service도 변경되어야 한다.
2. Response가 다른 방식으로 반환되어야 한다면? 마찬가지로 Service도 변경되어야 한다.

여러 Controller가 하나의 service를 사용하게 되는 상황에서
1. Controller마다 Request를 다르게 받는다면? Service 메서드도 Request에 따라 여러 개가 만들어져야 한다.
2. Controller가 각기 다른 Response를 요구한다면? 마찬가지로 Service 메서드도 Response에 따라 여러 개가 만들어져야 한다.

따라서, Service와 Controller 사이의 DTO도 분리하는 것이 적합하며, DTO의 위치는 "해당 DTO가 생성되는 위치"에 두어야 한다.

라는 결론이었습니다. ㅎㅎ

이렇게 되면, 만약 Request나 Response가 바뀌어도 
해당 Request, Response의 필드 + Request, Response <-> Service의 DTO 변환 로직
만 변경하면 되기 때문에 오히려 controller 입력값의 변경사항이 RequestDTO에만 영향을 미칠 것이라고 생각했어요!

물론 영이가 말씀해주신 것 처럼, 역 추가 등의 간단한 기능들 뿐인 현재 로직에서는 변경사항이 생길 가능성이 거의 없을 것 같고 Request의 필드 == DTO의 필드로 봐도 문제가 생기지는 않을 것 같습니다 😂

과한 분리인지, 아니면 여러 가능성을 생각했을 때 필요할지 고민이 많이 되는데 혹시 영이는 어떻게 생각하시나요 ?.?

choijy1705 replied:
저는 분리가 필요한 상황이 생길때 분리하는게 좋다고 생각해요!
미리 앞서서 분리하다보면 오히려 변경해야할 사항이 더 늘어날 가능성이 높은것 같아요!

choijy1705 commented on src/main/java/subway/application/dto/line/LineUpdateDto.java:
```diff
@@ -0,0 +1,25 @@
+package subway.application.dto.line;
+
+public class LineUpdateDto {
+    private Long id;
```
nullable하지 않다면 primitive 타입을 사용하면 어떨까요?

choijy1705 commented on src/main/java/subway/domain/PathFinder.java:
```diff
@@ -0,0 +1,64 @@
+        if (graph.containsVertex(source) && graph.containsVertex(dest)) {
+            return;
+        }
+        throw new StationNotFoundException();
```
에러 메세지도 적어주면 좋을것 같네요

somsom13 replied:
존재하지 않는 역이라는 것은 해당 예외에 대해서는 고정된 메세지 일 것 같아서 `~NotFoundException` 클래스의 내부에 넣어두었습니다.

반대로 `Illegal~Exception` 들은 상황에 따라 유동적으로 사용하기 위해 new로 예외 객체를 만들 때 메세지를 넣도록 설정했구요!

StationNotFoundException은 예외 클래스의 이름만 봐도 어떤 예외인지 알 수 있지 않을까~? 라고 생각했었는데요, 
그래도 통일성 + 예외가 발생하는 부분에서 메세지를 바로 볼 수 있다는 장점으로 에러 메세지를 외부에서 넣어주는게 더 좋을까요?

choijy1705 replied:
통일성 있게 외부에서 넣어주는것이 좋을것 같네요!

choijy1705 commented on src/main/java/subway/ui/LineController.java:
```diff
@@ -30,29 +31,29 @@ public LineController(LineService lineService) {
+        List<LineResponse> lineResponses = lineService.findLines()
+                .stream()
+                .map(LineResponse::from)
+                .collect(toList());
```
dto 변환 로직은 메서드로 빼서 핵심로직만 보여지도록 하면 좋을것 같아요!

choijy1705 commented on src/main/java/subway/domain/Path.java:
```diff
@@ -0,0 +1,25 @@
+
+public class Path {
+    private final List<Station> stations;
+    private final double distance;
```
distance는 기능과 의미하는 바가 조금 많은것 같은데
원시값 포장을 해보면 어떨까요?

choijy1705 reviewed (changes requested):
안녕하세요 바론!
2단계 엄청 잘 구현해주셨네요👍
핵심 기능들은 바로 머지해도 될것 같지만
마이너한 코멘트들이 많아서 다시한번 피드백 반영 요청드려요!

choijy1705 reviewed (changes requested):
안녕하세요 바론!
2단계 엄청 잘 구현해주셨네요👍
핵심 기능들은 바로 머지해도 될것 같지만
마이너한 코멘트들이 많아서 다시한번 피드백 반영 요청드려요!

somsom13 committed:
refactor: 사용되지 않는 코드 제거

somsom13 committed:
refactor: 경로 조회 API URI 변경

- /routes -> /paths

somsom13 committed:
refactor: 경로 조회 API 출발역, 도착역 데이터 변경

- RequestBody -> RequestParam
- 역 이름 -> 역 id

somsom13 committed:
refactor: Sections 정렬된 구간 조회 메서드 분리

- Sections: getSections(), getSectionsByList() 삭제, getSortedSections() 추가
- SectionRepository: 전체 구간 조회 메서드 findALl() 추가

somsom13 committed:
refactor: 불필요한 SectionEntity 조회 메서드 삭제

somsom13 committed:
feat: Distance VO 생성

somsom13 committed:
feat: Fare VO 생성

somsom13 committed:
refactor: profile 설정

somsom13 committed:
refactor: nullable 한 wrapper -> primitive 로 변경

somsom13 committed:
refactor: Line final 추가

somsom13 committed:
refactor: response 변환 메서드 분리

somsom13 committed:
refactor: FareCalculator util 클래스로 변경

somsom13 committed:
refactor: Exception 에외 메세지 밖에서 주입

somsom13 committed:
refactor: Request -> DTO 통합

somsom13 committed:
style: 구글 자바 컨벤션 적용

somsom13 commented:
안녕하세요 영이! 바론입니다 🙇‍♀️ 

어제 남겨주신 피드백들 마저 보고 반영했습니다 ㅎㅎ

리뷰 잘 부탁드립니다~!! 😸  감사합니다!! 🙇‍♀️

choijy1705 commented on src/main/java/subway/domain/Distance.java:
```diff
@@ -0,0 +1,64 @@
+        return Objects.hash(value);
+    }
+
+    public double value() {
+        return value;
+    }
```
getValue가 되야할것 같아요

choijy1705 commented on src/main/java/subway/domain/Distance.java:
```diff
@@ -0,0 +1,64 @@
+public class Distance {
+
+    private static final double MIN_VALUE = 0;
+    public static final Distance MIN_DISTANCE = new Distance(MIN_VALUE);
```
사용처가 FareCalculator 뿐인데
FareCalculator에서 관리되는게 더 좋지 않을까요?

choijy1705 commented on src/main/java/subway/ui/LineController.java:
```diff
@@ -29,30 +31,36 @@ public LineController(LineService lineService) {
-    @GetMapping("/detail")
-    public ResponseEntity<List<LineDetailResponse>> findAllDetailLines() {
-        return ResponseEntity.ok(lineService.findDetailLineResponses());
+    private List<LineResponse> mapToLineResponse(List<Line> lines) {
+        return lines.stream()
+            .map(LineResponse::from)
+            .collect(toList());
     }
```
핵심 로직이 아닌 것들은 하단으로 보내면 더 깔끔해보이겠네요!

choijy1705 commented on src/main/java/subway/ui/PathController.java:
```diff
@@ -0,0 +1,28 @@
+    }
+
+    @GetMapping
+    public ResponseEntity<ShortestPathResponse> findPath(@RequestParam Long sourceStationId,
+        @RequestParam Long destStationId) {
```
required option이 default가 true라서 null 체크가 되겠지만
명시적으로 primitive 타입인 long으로 처리해주는게 어떨까요?

choijy1705 commented on src/test/java/subway/dao/LineDaoTest.java:
```diff
@@ -11,13 +11,18 @@
 @JdbcTest
+@AutoConfigureTestDatabase(replace = Replace.NONE)
 @Sql("/line_initialize.sql")
+@ActiveProfiles("test")
```
src/test/resources 에 application.yml 만들고
```yml
spring:
  profiles:
    active: test
```
이렇게 하거나 build.gradle에

```
test {
	systemProperty "spring.profiles.active", "test"
	useJUnitPlatform()
}

```
이렇게 추가해주면 매번 activeProfiles를 설정하지 않아도 될것 같아요

choijy1705 reviewed (approved):
안녕하세요 바론!
피드백 반영 정말 잘해주셨네요!👍
간단한 코멘트 남겼으니 확인해주세요
이번 미션은 이만 머지하겠습니다
